clib/polyMap.ml

```diff
@@ -0,0 +1,89 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+module type ValueS =
+sig
+  type 'a t
+end
+
+module type Tag = sig type _ tag = .. end
+
+module Make (Tag:Tag) = struct
+  open Tag
+
+  module type OneTag = sig
+    type a
+    type _ tag += T : a tag
+  end
+  type 'a onetag = (module OneTag with type a = 'a)
+
+  let refl = Some CSig.Refl
+
+  let eq_onetag (type a b) (tag:a onetag) (tag':b tag) : (a,b) CSig.eq option =
+    let module T = (val tag) in
+    match tag' with
+    | T.T -> refl
+    | _ -> None
+
+  let make (type a) () : a onetag =
+    (module struct type nonrec a = a type _ tag += T : a tag end)
+
+  let tag_of_onetag (type a) (tag:a onetag) : a tag =
+    let module T = (val tag) in
+    T.T
+
+  module type MapS = sig
+    type t
+    type _ value
+
+    val empty : t
+
+    val find : 'a tag -> t -> 'a value
+
+    val add : 'a onetag -> 'a value -> t -> t
+
+    val mem : 'a tag -> t -> bool
+
+    val modify : 'a tag -> ('a value -> 'a value) -> t -> t
+
+  end
+
+  module Map (V:ValueS) = struct
+
+    type v = V : 'a onetag * 'a V.t -> v
+
+    let key t = Obj.Extension_constructor.(id (of_val t))
+    let onekey t = key (tag_of_onetag t)
+
+    module M = Int.Map
```

Comment:
Using Int.Map on the extension constructor feels cleaner than HMap using polymorphic compare and hash